python-package/lightgbm/basic.py

```diff
@@ -237,45 +237,60 @@ def _data_to_2d_numpy(data: Any, dtype: type = np.float32, name: str = 'list') -
                     "It should be list of lists, numpy 2-D array or pandas DataFrame")
 
 
-def cfloat32_array_to_numpy(cptr, length):
+def cfloat32_array_to_numpy(cptr: ctypes.POINTER, length: int) -> np.ndarray:
     """Convert a ctypes float pointer array to a numpy array."""
     if isinstance(cptr, ctypes.POINTER(ctypes.c_float)):
         return np.ctypeslib.as_array(cptr, shape=(length,)).copy()
     else:
         raise RuntimeError('Expected float pointer')
 
 
-def cfloat64_array_to_numpy(cptr, length):
+def cfloat64_array_to_numpy(cptr: ctypes.POINTER, length: int) -> np.ndarray:
     """Convert a ctypes double pointer array to a numpy array."""
     if isinstance(cptr, ctypes.POINTER(ctypes.c_double)):
         return np.ctypeslib.as_array(cptr, shape=(length,)).copy()
     else:
         raise RuntimeError('Expected double pointer')
 
 
-def cint32_array_to_numpy(cptr, length):
+def cint32_array_to_numpy(cptr: ctypes.POINTER, length: int) -> np.ndarray:
     """Convert a ctypes int pointer array to a numpy array."""
     if isinstance(cptr, ctypes.POINTER(ctypes.c_int32)):
         return np.ctypeslib.as_array(cptr, shape=(length,)).copy()
     else:
         raise RuntimeError('Expected int32 pointer')
 
 
-def cint64_array_to_numpy(cptr, length):
+def cint64_array_to_numpy(cptr: ctypes.POINTER, length: int) -> np.ndarray:
     """Convert a ctypes int pointer array to a numpy array."""
     if isinstance(cptr, ctypes.POINTER(ctypes.c_int64)):
         return np.ctypeslib.as_array(cptr, shape=(length,)).copy()
     else:
         raise RuntimeError('Expected int64 pointer')
 
 
-def c_str(string):
+def c_str(string: str) -> ctypes.c_char_p:
     """Convert a Python string to C string."""
     return ctypes.c_char_p(string.encode('utf-8'))
 
 
-def c_array(ctype, values):
-    """Convert a Python array to C array."""
+def c_array(ctype:Union[ctypes.c_char_p,ctypes.c_int], values:List[Union[str,int]])->np.ndarray:
```

Comment:
```suggestion
def c_array(ctype: type, values: List[ctypes.c_char_p]) -> ctypes.Array:
```

This function is only called in one place in `lightgbm`.

https://github.com/microsoft/LightGBM/blob/273c9b043e11b24a95e40f99ef81c4888d2089c9/python-package/lightgbm/basic.py#L2178

Found with the following

```shell
git grep -E 'c_array\('
```

In that place, it's called with type `ctypes.c_char_p` (not an *instance* of that type) and provided a list of `ctypes.c_char_p` instances.

Here's some code I used to figure out the return type of this function.

```python
import ctypes

values = [
    ctypes.c_char_p("a".encode('utf-8')),
    ctypes.c_char_p("b".encode('utf-8')),
    ctypes.c_char_p("c".encode('utf-8'))
]

x = (ctypes.c_char_p * len(values))(*values)
print(type(x).__mro___)
isinstance(x, ctypes.Array)
```

Please make these changes, and for the purpose of this PR please revert the changes you've made to the function's docstring. Those changes contain a few inaccurate statements, and are not necessary for the scope of this PR.

Reply:
You're right, I made a similar test, however wasn't sure if `ctype.Array` was the right one of typing. 
I believe this was the last revision issue. Let me know if tou see something unusual.